doc/sphinx/proof-engine/tactics.rst

```diff
@@ -105,6 +108,354 @@ bindings_list`` where ``bindings_list`` may be of two different forms:
+   while :n:`and` has a single constructor (:n:`conj`) with multiple parameters
+   (:n:`A` and :n:`B`).
+   These are defined in theories/Init/Logic.v.  The "where" clauses define the
+   infix notation for "or" and "and".
+
+   | Inductive or (A B:Prop) : Prop :=
+   |   | or_introl : A -> A \\/ B
+   |   | or_intror : B -> A \\/ B
+   | where "A \\/ B" := (or A B) : type_scope.
+   |
+   | Inductive and (A B:Prop) : Prop :=
+   |   conj : A -> B -> A /\\ B
+   | where "A /\\ B" := (and A B) : type_scope.
+
+   **quoted block above isn't formatted as I wanted.  What's the right way to get output to exactly match the input?**
+
+.. note::
+
+   :n:`intros {+ p}` is not always equivalent to :n:`intros p; ... ; intros p`
+   if some of the :n:`p` is :g:`_`.  In the first form, all erasures are done
+   at once, while they're done sequentially for each tactic in the second form.
+   If the second matched term depends on the first matched term and the pattern
+   for both is :g:`_` (i.e., both will be erased), the first :n:`intros` in the second
+   form will fail because the second matched term still has the dependency on the first.
+
+Examples:
+
+.. _intropattern_conj_ex:
+
+   .. example:: intro pattern for /\\
+
+      .. coqtop:: reset none
+
+         Goal forall (A: Prop) (B: Prop), (A /\ B) -> True.
+
+      .. coqtop:: out
+
+         intros.
+
+      .. coqtop:: all
+
+         destruct H as (HA & HB).
+
+.. _intropattern_disj_ex:
+
+   .. example:: intro pattern for \\/
+
+      .. coqtop:: reset none
+
+         Goal forall (A: Prop) (B: Prop), (A \/ B) -> True.
+
+      .. coqtop:: out
+
+         intros.
+
+      .. coqtop:: all
+
+         destruct H as [HA|HB]. all: swap 1 2.
+
+.. _intropattern_rarrow_ex:
+
+   .. example:: -> intro pattern
+
+      .. coqtop:: reset none
+
+         Goal forall (x:nat) (y:nat) (z:nat), (x = y) -> (y = z) -> (x = z).
+
+      .. coqtop:: out
+
+         intros * H.
+
+      .. coqtop:: all
+
+         intros ->.
+
+.. _intropattern_inj_discr_ex:
+
+   .. example:: [=] intro pattern
+
+      The first :n:`intros [=]` uses :tacn:`injection` to strip :n:`(S ...)` from
+      both sides of the matched equality.  The second uses :tacn:`discriminate` on
+      the contradiction :n:`1 = 2` (internally represented as :n:`(S O) = (S (S O))`)
+      to complete the goal.
+
+      .. coqtop:: reset none
+
+         Goal forall (n m:nat),  (S n) = (S m) -> (S O)=(S (S O)) -> False.
+
+      .. coqtop:: out
+
+         intros *.
+
+      .. coqtop:: all
+
+         intros [= H].
+
+      .. coqtop:: all
+
+         intros [=].
+
+.. _intropattern_injection_ex:
+
+   .. example:: [=->%length_zero_iff_nil] intro pattern
+
+      .. from https://github.com/tchajed/coq-tricks/blob/master/src/IntroPatterns.v
+
+      | :g:`Theorem length_zero_iff_nil (l : list A):`
+      |    :g:`length l = 0 <-> l=[].`
+
+      * `intros` would add :g:`H : S (length ys) = 1`
```

Comment:
The ordering is weird. It would be natural if this were commenting on the code above but this is commenting on the text below. What about reversing completely the order in this example: first the code, then the bullet points, then the theorem statement.